src/algebra/covariant_and_contravariant.lean

```diff
@@ -97,9 +95,166 @@ for all `m ∈ M` and all elements `n₁, n₂ ∈ N`, if the relation `r` holds
+    exact h a⁻¹ bc },
+  { rw [← inv_mul_cancel_left a b, ← inv_mul_cancel_left a c] at bc,
+    exact h a⁻¹ bc }
+end
+
+lemma covconv [group N] [cov : covariant_class N N (*) r] : contravariant_class N N (*) r :=
+{ covtc := λ a b c bc, group.covariant_iff_contravariant.mp cov.covc _ bc }
+
+
+section is_trans
+variables [is_trans N r] (m n : M) {a b c d : N}
+
+/-  Lemmas with 3 elements. -/
+lemma act_rel_of_rel_of_act_rel (ab : r a b) (rl : r (μ m b) c) :
+  r (μ m a) c :=
+trans (act_rel_act_of_rel m ab) rl
+
+lemma rel_act_of_rel_of_rel_act (ab : r a b) (rr : r c (μ m a)) :
+  r c (μ m b) :=
+trans rr (act_rel_act_of_rel _ ab)
+
+end is_trans
+
+end covariant
+
+/-  Lemma with 4 elements. -/
+section M_eq_N
+variables {M N μ r} {mu : N → N → N} [is_trans N r]
+  [covariant_class N N mu r] [covariant_class N N (function.swap mu) r] {a b c d : N}
+
+lemma act_rel_act_of_rel_of_rel (ab : r a b) (cd : r c d) :
+  r (mu a c) (mu b d) :=
+trans (act_rel_act_of_rel c ab : _) (act_rel_act_of_rel b cd)
+
+end M_eq_N
+
+section contravariant
+variables {M N μ r} [contravariant_class M N μ r]
+
+lemma rel_of_act_rel_act (m : M) {a b : N} (ab : r (μ m a) (μ m b)) :
+  r a b :=
+contravariant_class.covtc _ ab
+
+section is_trans
+variables [is_trans N r] (m n : M) {a b c d : N}
+
+/-  Lemmas with 3 elements. -/
+lemma act_rel_of_act_rel_of_rel_act_rel (ab : r (μ m a) b) (rl : r (μ m b) (μ m c)) :
+  r (μ m a) c :=
+trans ab (rel_of_act_rel_act m rl)
+
+lemma rel_act_of_act_rel_act_of_rel_act (ab : r (μ m a) (μ m b)) (rr : r b (μ m c)) :
+  r a (μ m c) :=
+trans (rel_of_act_rel_act m ab) rr
+
+end is_trans
+
+end contravariant
+
+lemma covariant_le_of_covariant_lt [partial_order N] :
+  covariant M N μ (<) → covariant M N μ (≤) :=
+begin
+  refine λ h a b c bc, _,
+  rcases le_iff_eq_or_lt.mp bc with rfl | bc,
+  { exact rfl.le },
+  { exact (h _ bc).le }
+end
+
+lemma contravariant_lt_of_contravariant_le [partial_order N] :
+  contravariant M N μ (≤) → contravariant M N μ (<) :=
+begin
+  refine λ h a b c bc, lt_iff_le_and_ne.mpr ⟨h a bc.le, _⟩,
+  rintro rfl,
+  exact lt_irrefl _ bc,
+end
+
+lemma covariant_le_iff_contravariant_lt [linear_order N] :
+  covariant M N μ (≤) ↔ contravariant M N μ (<) :=
+⟨ λ h a b c bc, not_le.mp (λ k, not_le.mpr bc (h _ k)),
+  λ h a b c bc, not_lt.mp (λ k, not_lt.mpr bc (h _ k))⟩
+
+lemma covariant_lt_iff_contravariant_le [linear_order N] :
+  covariant M N μ (<) ↔ contravariant M N μ (≤) :=
+⟨ λ h a b c bc, not_lt.mp (λ k, not_lt.mpr bc (h _ k)),
+  λ h a b c bc, not_le.mp (λ k, not_le.mpr bc (h _ k))⟩
+
+@[to_additive]
+lemma covariant_iff_covariant_mul [comm_semigroup N] :
+  covariant N N (*) (r) ↔ covariant N N (flip (*)) (r) :=
+by rw is_symm_op.flip_eq
+
+@[to_additive]
+lemma contravariant_mul_iff_flip [comm_semigroup N] :
+  contravariant N N (*) (r) ↔ contravariant N N (flip (*)) (r) :=
+by rw is_symm_op.flip_eq
+
+@[to_additive]
+instance covariant_mul_le.to_contravariant_lt_mul [has_mul N] [linear_order N]
```

Comment:
inconsistent name (what is the order of operation and relation in names?), also projection notation will not work here.
`contravariant_mul_lt_of_covariant_mul_le`? (same below a couple of times)

Reply:
Adopted your names!  I am always confused by the ordering of stuff...